### PR TITLE
[Bug] Update Binding moves to affect Ghosts with damage

### PR DESCRIPTION
**Pull request by Nohbody**
## What are the changes?
1. Fixes Damaging Trap moves to apply the damaging effect while still allowing them to switch out.

## Why am I doing these changes?
1. Fixes: https://github.com/pagefaultgames/pokerogue/issues/1915

## What did change?
1. Binding moves as listed here: https://bulbapedia.bulbagarden.net/wiki/Bound#Binding_moves can be applied to Ghost types in order to get the Damage over Time effects (e.g. 1/8 max hp at the end of the turn).

## How to test the changes?
1. Set your overrides.ts:
```
export const MOVESET_OVERRIDE: Array<Moves> = [Moves.GROWL, Moves.PROTECT];
export const OPP_MOVESET_OVERRIDE: Array<Moves> = [Moves.INFESTATION];
```
2. Get into a battle with 2 non-Ghost Pokemon and a Ghost Pokemon.
3. Use Growl and let the opponent hit you with the Binding move (Infestation) on a non-Ghost.
4. Verify with Protect that you are taking the DoT damage.
5. Try to switch out to a different Pokemon (it should not let you).
6. When the Binding move ends or you faint, switch to the Ghost Pokemon.
7. Use Growl and let the opponent hit you with the Binding move (Infestation).
8. Verify with Protect that you are taking the DoT damage. (**new**)
9. Try to switch out to a different Pokemon (it **should** let you). (**new**)

Verify this is working against an Opponent Trainer with a Ghost-type or not (unsure how to force this).

## Checklist
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [ ] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/data/battler-tags.ts [hunk 1]
```diff
@@ -126,7 +126,9 @@ export class TrappedTag extends BattlerTag {
   onAdd(pokemon: Pokemon): void {
     super.onAdd(pokemon);
 
-    pokemon.scene.queueMessage(this.getTrapMessage(pokemon));
+    if (!pokemon.isOfType(Type.GHOST)) {
+      pokemon.scene.queueMessage(this.getTrapMessage(pokemon));
+    }
```
Comment (Nohbody) [inline src/data/battler-tags.ts:131]: This prevents the "no longer can escape!" message from appearing if used on a Ghost Pokemon.

File: src/data/battler-tags.ts [hunk 2]
```diff
@@ -689,7 +691,7 @@ export abstract class DamagingTrapTag extends TrappedTag {
   }
 
   canAdd(pokemon: Pokemon): boolean {
-    return !pokemon.isOfType(Type.GHOST) && !pokemon.findTag(t => t instanceof DamagingTrapTag);
+    return !pokemon.findTag(t => t instanceof DamagingTrapTag);
```
Comment (Nohbody) [inline src/data/battler-tags.ts:694]: Damaging Trap (Binding) moves can not stack DoT effects.

**Comment by Nohbody** [comment]
@Madmadness65 Can this be looked at? It's starting to become stale.

**Comment by Nohbody** [comment]
@Tempo-anon Not sure who should review this.

File: src/phases.ts [hunk 3]
```diff
@@ -2075,7 +2075,9 @@ export class CommandPhase extends FieldPhase {
             this.scene.currentBattle.turnCommands[this.fieldIndex - 1].skip = true;
           }
         } else if (trapTag) {
-          if (trapTag.sourceMove === Moves.INGRAIN && this.scene.getPokemonById(trapTag.sourceId).isOfType(Type.GHOST)) {
+          const canSwitch = (trapTag.sourceMove === Moves.INGRAIN && this.scene.getPokemonById(trapTag.sourceId).isOfType(Type.GHOST)) ||
+                            (trapTag instanceof DamagingTrapTag && playerPokemon.isOfType(Type.GHOST));
```
Comment (f-fsantos) [inline src/phases.ts:2079]: Why are you not using the same getPokemonById method as above?

Comment (Nohbody) [reply src/phases.ts:2079]: I can't recall if there's any reason why they wouldn't resolve the same as `playerPokemon` anyways.

**Comment by DayKev** [comment]
Could you update the PR so it can be [re-]evaluated?

**Comment by Snailman11** [comment]
Link to Discord's bug report (ghost types dont take the dot from infestation)
https://discord.com/channels/1125469663833370665/1254970595465232394

**Comment by DayKev** [comment]
Sure, gotta fix some of my PRs first then I'll move to this one.

**Comment by DayKev** [comment]
Reopening so GH Desktop will see the PR and allow me to pull it.

**Pull request closed without merging**